Faster packing

austinpagan opened the pull request:
# Description

After some review, I realized that the eight packing routines in my code for Power-optimized 8-bit matrix multiplication were slow (and also fairly hard to read and understand).  
I've replaced them with new routines.
The four 8-bit packing routines now run 4x faster than before.
The four 16-bit "fallback" packing routines run 2x faster than previously.
The code size for these eight routines is roughly half of what it was before.

Fixes # (github issue)

# Checklist

## General

- [ YES] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit?
- [ YES] Have you formatted the code using clang-format?

## Performance improvements

- [ See answer] Have you submitted performance data that demonstrates performance improvements?
   I will attach a spreadsheet in a comment below that summarizes the improvements I saw.

### New features

- [ n/a] Have you published an RFC for the new feature?
- [ n/a] Was the RFC approved?
- [ n/a] Have you added relevant tests?

### Bug fixes

- [ n/a] Have you included information on how to reproduce the issue (either in a github issue or in this PR)?
- [ n/a] Have you added relevant regression tests?

## [RFC](https://github.com/oneapi-src/oneDNN/tree/rfcs) PR

- [ n/a] Does RFC document follow the [template](https://github.com/oneapi-src/oneDNN/blob/rfcs/rfcs/template.md#onednn-design-document-rfc)?
- [ n/a] Have you added a link to the rendered document?

austinpagan commented:
Here's a spreadsheet showing improvements, run on four matrix types: 64x64, 128x128, 256x256 and 512x512.
In all four cases I checked performance in my 8-bit code (no pesky offsets) and my 16-bit code (fallback to handle offset situations):

[Summary Packing Improvement.xlsx](https://github.com/oneapi-src/oneDNN/files/9603664/Summary.Packing.Improvement.xlsx)

austinpagan commented:
@dzarukin Happy to address any questions you might have.

mgouicem commented on src/cpu/ppc64/ppc64_gemm_s8x8s32.hpp:
```diff
@@ -20,2108 +20,1255 @@
-    }
+static inline void memcpy_16(void * __restrict d, const void * __restrict s) {
+        int *di, *si;
+        di = (int *)(d);
```
(nit) use int32_t to specify the type size?
Also, compilers have optimized/vectorize implementations for memcpy. [Here ](https://godbolt.org/z/TeMPrvo1f) is an example on x64, where a call to memcpy with fixed size are just swapped with a single instruction.

Alternatively, I think 4 calls to memcpy_4 should produce the same code for most compilers. [Here ](https://godbolt.org/z/d9WEbTE1P) is an example.

mgouicem commented on src/cpu/ppc64/ppc64_gemm_s8x8s32.hpp:
```diff
@@ -20,2108 +20,1255 @@
-        _di[2] = _si[2]; \
-        _di[3] = _si[3]; \
-    }
+static inline void memcpy_14(void * __restrict d, const void * __restrict s) {
```
this function seems unused.

mgouicem reviewed (commented):
This looks good to me, thanks.
Just two comments overall:
-  Could you please specify ppc64 in the commit messages please.  Something like "cpu: ppc64: commit message"?
- why implement all the routines in a header file?

austinpagan commented:
Sorry for delay in response.  I'm on vacation through the next week, and will address the changes you requested at that time.

As far as the huge header file, here was my thought process:
The C file was to hold the "transition code" which handled the offset stuff and the decision on whether we would fall back to using 16-bit rather than 8-bit operations. The big header file was to cleanly contain just the eight packing routines and the two kernels.
If this is inappropriate, I'd invite you to suggest how you'd have me do it instead.

austinpagan commented:
So, I changed all "int" to "int32_t", and removed all the unused memcpy stuff (my efforts to shrink the packing code had removed the calls, I just needed to remove the definitions).  The "third" fix was to mention "ppc64" in the commit comment!

austinpagan commented:
The one failing build above is failing the three "ACL" builds, and the one failing test is called
`test_convolution_eltwise_forward_f32`
My code only runs on PPC64 boxes, and it is only brought in when we're doing 8-bit matrix multiplication.
I'm thinking it's unlikely that this pull request is the cause of this error?

nSircombe commented:
Hi @austinpagan,
The Drone CI pipeline appears to be failing due to the recent API changes, see https://github.com/oneapi-src/oneDNN/pull/1440#issuecomment-1267747939, which weren't comparable with ACL.

@dzarukin was looking at a fix, which I think should go in sometime soon.

austinpagan commented:
@dzarukin:
I hope you're having good fortune in dealing with the error situation mentioned above.
I'm looking for some rough estimate when this PR might be promoted, please?

austinpagan commented:
@dzarukin -- Happy Hallowe'en...  here's your spooky reminder that I'm still hoping for acceptance of this pull request?

dzarukin commented:
> @dzarukin -- Happy Hallowe'en... here's your spooky reminder that I'm still hoping for acceptance of this pull request?

Hi @austinpagan, yeah, sorry about it, it fell off our radar for a while. We'll promote changes soon. Thank you for waiting.

dzarukin commented:
Sorry for a delay and thank you for the contribution!
